### ISSUE TITLE
Log-Likelihood or Kullback Leibler-Divergence - which is best for STIR?

### ISSUE THREAD

**Issue by robbietuk** [UCL/STIR#697]
I have investigating the use of the Kullback Leibler-Divergence (KL) in place of the Log-Likelihood (LL) in STIR. The two functions both have the same shape, but with a different offset. 

Log-Likelihood
![image](https://user-images.githubusercontent.com/13292409/94438287-d2e6df00-0196-11eb-9dea-956039086bad.png)

KL
![image](https://user-images.githubusercontent.com/13292409/94438318-dda17400-0196-11eb-8c61-c635c28b2dac.png)

where \bar{y} is the expected data and y the measured. If found that KL generally results in smaller values than LL, which can be up to 3 orders of magnetude and thus, evaluation of the data fitting term less susseptable to numerical error, although I am unsure as to the current impact of numerical error. In #528 the numerical precision was improved by computing the log-likelihood as doubles and since I have not observed such issues.

**Comment by robbietuk** [comment]
The offset the log-likelihood is causing more problems convergence computation. It was suggested to use a bool variable, such as `use_negitive_KL`, in the `GeneralisedObjectiveFunction` to allow for the computation objective function value to use the negitive KL divergence instead of LL.

**Comment by robbietuk** [comment]
We should have some method that allows for the toggling between the two data fit values. I expect this to be a property of `GeneralisedObjectiveFunction` but the computation takes place here here:https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/buildblock/recon_array_functions.cxx#L347-L395 
but there is also a seperate computation here:
https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/buildblock/recon_array_functions.cxx#L183-L277

**Comment by KrisThielemans** [comment]
`divide_and_truncate` is used in the gradient calculation. 
https://github.com/UCL/STIR/blob/b79bb323c1f1abab02f397f04ea8a13995b133c2/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L1079
It has a feature that optionally it can also compute the objective function, saving you having to do another forward projection. However, of course it only computes it for the current subset.

It turns out that we currently don't use this facility. (it is available in [distributable_compute_gradient](https://github.com/UCL/STIR/blob/b79bb323c1f1abab02f397f04ea8a13995b133c2/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L921-L933), but isn't used where that is called in [compute_sub_gradient_without_penalty_plus_sensitivity](https://github.com/UCL/STIR/blob/b79bb323c1f1abab02f397f04ea8a13995b133c2/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L620-L640) which in some situations is a pity (e.g. when doing line-searches).

In any case, we need to have those 2 calculations to be consistent with each-other.

So it seems we need a switch in `PoissonLogLikelihoodWithLinearModelForMean`, and then pass that to the 2 functions you found above (as an extra parameter with default "do the old thing").

**Comment by robbietuk** [comment]
It appears that these two calculations are not consistant right now. The `divide_and_truncate` method does not subtract the `denom`/`estimated_data` term.
https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/buildblock/recon_array_functions.cxx#L253-L267
It is therefore only computing `y log(ybar)` and not `y log(ybar) - ybar` (obviously reading into this a little as there are divide by 0 corrections here). 

I am working on a PR (70%) done.

**Comment by KrisThielemans** [comment]
hmm. I think you are right. I guess need to check more carefully where it is used (but probably that part is effectively not used at all at present).

**Comment by robbietuk** [comment]
I dont think it is really, 2 uses, both called through `divide_and_truncate` for RelatedViewgrams inputs. https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/buildblock/recon_array_functions.cxx#L281

1. `RPC_process_related_viewgrams_gradient` https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L1031-L1082
This can be traced back to `distributable_compute_gradient` where `log_likelihood_ptr` is setup as `NULL`. 

2. `actual_add_multiplication_with_approximate_sub_Hessian_without_penalty`
https://github.com/UCL/STIR/blob/79031e4f72f4a6407b2a67dfecd50aad7584e609/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L809-L893
which does not setup a `log_likelihood_ptr` so it never uses the computation, I think this means it is `NULL`.

I think this means I can fix this without breaking backward compatiblity.

**Comment by KrisThielemans** [comment]
ok
